src/storage/src/hummock/tiered_cache.rs

```diff
@@ -0,0 +1,200 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::hash::{BuildHasher, Hash};
+
+pub trait HashBuilder = BuildHasher + Clone + Send + Sync + 'static;
+
+pub trait TieredCacheKey: Eq + Send + Sync + Hash + Clone + 'static + std::fmt::Debug {
+    fn encoded_len() -> usize;
+
+    fn encode(&self, buf: &mut [u8]);
+
+    fn decode(buf: &[u8]) -> Self;
+}
+
+#[expect(clippy::len_without_is_empty)]
+pub trait TieredCacheValue: Send + Sync + Clone + 'static {
+    fn len(&self) -> usize;
+
+    fn encoded_len(&self) -> usize;
+
+    fn encode(&self, buf: &mut [u8]);
+
+    fn decode(buf: Vec<u8>) -> Self;
+}
+
+pub enum TieredCacheEntry<K, V>
+where
+    K: TieredCacheKey,
+    V: TieredCacheValue,
+{
+    Cache(CachableEntry<K, V>),
+    Owned(Box<V>),
+}
+
+pub struct TieredCacheEntryHolder<K, V>
+where
+    K: TieredCacheKey,
+    V: TieredCacheValue,
+{
+    handle: TieredCacheEntry<K, V>,
+    value: *const V,
+}
+
+impl<K, V> TieredCacheEntryHolder<K, V>
+where
+    K: TieredCacheKey,
+    V: TieredCacheValue,
+{
+    pub fn from_cached_value(entry: CachableEntry<K, V>) -> Self {
+        let ptr = entry.value() as *const _;
+        Self {
+            handle: TieredCacheEntry::Cache(entry),
+            value: ptr,
+        }
+    }
+
+    pub fn from_owned_value(value: V) -> Self {
+        let value = Box::new(value);
+        let ptr = value.as_ref() as *const _;
+        Self {
+            handle: TieredCacheEntry::Owned(value),
+            value: ptr,
+        }
+    }
+
+    pub fn into_inner(self) -> TieredCacheEntry<K, V> {
+        self.handle
+    }
+
+    pub fn into_owned(self) -> V {
+        match self.handle {
+            // TODO(MrCroxx): There is a copy here, eliminate it by erase the entry later.
+            TieredCacheEntry::Cache(entry) => entry.value().clone(),
+            TieredCacheEntry::Owned(value) => *value,
+        }
+    }
+}
+
+impl<K, V> std::ops::Deref for TieredCacheEntryHolder<K, V>
+where
+    K: TieredCacheKey,
+    V: TieredCacheValue,
+{
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &(*self.value) }
+    }
+}
+
+use std::marker::PhantomData;
+
+use risingwave_common::cache::CachableEntry;
+
+#[cfg(target_os = "linux")]
+pub use super::file_cache;
+
+#[derive(thiserror::Error, Debug)]
+pub enum TieredCacheError {
+    #[cfg(target_os = "linux")]
+    #[error("file cache error: {0}")]
+    FileCache(#[from] file_cache::error::Error),
+}
+
+pub type Result<T> = core::result::Result<T, TieredCacheError>;
+
+pub enum TieredCacheOptions {
+    NoneCache,
+    #[cfg(target_os = "linux")]
+    FileCache(file_cache::cache::FileCacheOptions),
+}
+
+pub enum TieredCache<K, V>
```

Comment:
What will this enum be used? Will NoneCache only be used in non-linux target? If yes, I wonder whether we can prevent creating a TieredCache for non-linux target instead of introducing an enum, which makes each interface comes with a `match`.

Reply:
`NoneCache` can be used on any target, but `FileCache` only supports linux target.

Introducing `TieredCache` enum is good for integrating tiered cache into hummock, or we must add `#[cfg(target_os = "target")]` everywhere we use file cache in hummock code.

Reply:
I was thinking about introducing a unify cache interface for both memory cache and file cache. In this case, if the target os linux and file cached is enabled, we create a cache instance containing both memory cache and file cache (similar to what we did in HybridObjectStore). Otherwise, we only create a block cache instance. I don't have a strong opinion on this so we can revisit it when we have file cache fully integrated with the memory cache.

Reply:
I'm not quite sure if unifying memory cache and tiered cache is a good idea, 🤔. Memory cache call are synchronous but tiered cache calls are asynchronous, unifying them may introduce extra cost for memory cache. Besides (weakly), `#[async_trait]` is not zero-cost yet, we need to use `GAT` to reduce the boxing cost (which is not that friendly for everyone, but affordable).

Reply:
> I'm not quite sure if unifying memory cache and tiered cache is a good idea, 🤔. Memory cache call are synchronous but tiered cache calls are asynchronous, unifying them may introduce extra cost for memory cache. Besides (weakly), `#[async_trait]` is not zero-cost yet, we need to use `GAT` to reduce the boxing cost (which is not that friendly for everyone, but affordable).

Ah... I missed that file cache has async interface. Just ignore what i just said.